In [61]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
%matplotlib inline

In [62]:
os.rename("WELLCOME_APCspend2013_forThinkful.csv", "APCspend2013")

FileNotFoundError: [WinError 2] The system cannot find the file specified: 'WELLCOME_APCspend2013_forThinkful.csv' -> 'APCspend2013'

In [63]:
df = pd.read_csv("APCspend2013", encoding = "ISO-8859-1")

In [66]:
df.head(n=20)

PMID/PMCID  \
0                            NaN   
1                     PMC3679557   
2          23043264  PMC3506128    
3            23438330 PMC3646402   
4           23438216 PMC3601604    
5                     PMC3579457   
6                     PMC3709265   
7           23057412 PMC3495574    
8              PMCID: PMC3780468   
9              PMCID: PMC3621575   
10             PMCID: PMC3739413   
11             PMCID: PMC3530961   
12             PMCID: PMC3624797   
13                    PMC3413243   
14                    PMC3694353   
15                    PMC3572711   
16                      22610094   
17             PMCID: PMC3586974   
18  23455506  PMCID: PMC3607399    
19    PMID: 24015914 PMC3833349    

                                       Publisher  \
0                                            CUP   
1                                            ACS   
2                                            ACS   
3                                            ACS   
4                                            ACS   
5                                            ACS   
6                                            ACS   
7                                            ACS   
8   ACS (Amercian Chemical Society) Publications   
9   ACS (Amercian Chemical Society) Publications   
10  ACS (Amercian Chemical Society) Publications   
11  ACS (Amercian Chemical Society) Publications   
12  ACS (Amercian Chemical Society) Publications   
13                              ACS Publications   
14                              ACS Publications   
15                                 AGA Institute   
16                                         AMBSB   
17         American Association of Immunologists   
18         American Association of Immunologists   
19                     American Chemical Society   

                                   Journal title  \
0                         psychological medicine   
1                              biomacromolecules   
2                                     j med chem   
3                                     j med chem   
4                                     j org chem   
5                 journal of medicinal chemistry   
6                   journal of proteome research   
7                                      mol pharm   
8                           acs chemical biology   
9                           acs chemical biology   
10  journal of chemical information and modeling   
11                journal of medicinal chemistry   
12                journal of medicinal chemistry   
13                                  biochemistry   
14                journal of medicinal chemistry   
15                              gastroenterology   
16               journal of biological chemistry   
17                         journal of immunology   
18                         journal of immunology   
19                          acs chemical biology   

                                        Article title  \
0   Reduced parahippocampal cortical thickness in ...   
1   Structural characterization of a Model Gram-ne...   
2   Fumaroylamino-4,5-epoxymorphinans and related ...   
3   Orvinols with mixed kappa/mu opioid receptor a...   
4   Regioselective opening of myo-inositol orthoes...   
5   Comparative Structural and Functional Studies ...   
6   Mapping Proteolytic Processing in the Secretom...   
7   Quantitative silencing of EGFP reporter gene b...   
8   A Novel Allosteric Inhibitor of the Uridine Di...   
9   Chemical proteomic analysis reveals the drugab...   
10  Locating Sweet Spots for Screening Hits and Ev...   
11  Synthesis and evaluation of ?-thymidine analog...   
12  Discovery and structure-activity relationships...   
13  Monomeric 14-3-3 has a chaperone-like activity...   
14  Synthesis and in vitro and in vivo characteriz...   
15  Much of the genetic risk of colorectal cancer ...   
16                Annexin-1 interaction with FPR2/ALX   
17  Sialoadhesin ligand expression identifies a su...

**Directions**: Determine the five most common journals and the total articles for each. Next, calculate the mean, median, and standard deviation of the open-access cost per article for each journal. For a real bonus round, identify the open access prices paid by subject area.

In [65]:
df["Journal title"] = df["Journal title"].str.lower()

In [18]:
df["Journal title"].value_counts()

plos one                                               190
journal of biological chemistry                         53
neuroimage                                              29
plos pathogens                                          24
plos genetics                                           24
nucleic acids research                                  23
plos neglected tropical diseases                        20
proceedings of the national academy of sciences         20
nature communications                                   19
human molecular genetics                                19
brain                                                   14
bmc public health                                       14
movement disorders                                      13
journal of neuroscience                                 12
developmental cell                                      12
biochemical journal                                     12
journal of general virology                             

ANSWER: The five most common journals are...
1. Plos One (190 articles)
2. Journal of Biological Chemistry (53 articles)
3. Neuroimage (29 articles)
4. Plos Pathogens (24 articles)
5. Plos Genetics (24 articles)

In [19]:
df.dtypes

PMID/PMCID                                             object
Publisher                                              object
Journal title                                          object
Article title                                          object
COST (£) charged to Wellcome (inc VAT when charged)    object
dtype: object

In [32]:
df["COST (£) charged to Wellcome (inc VAT when charged)"] = df["COST (£) charged to Wellcome (inc VAT when charged)"].str.replace("£", "")

In [34]:
df["COST (£) charged to Wellcome (inc VAT when charged)"] = df["COST (£) charged to Wellcome (inc VAT when charged)"].str.replace("$", "")

In [37]:
df["COST (£) charged to Wellcome (inc VAT when charged)"] = pd.to_numeric(df["COST (£) charged to Wellcome (inc VAT when charged)"])

In [38]:
df["COST (£) charged to Wellcome (inc VAT when charged)"].mean()

24067.33997179131

In [39]:
df["COST (£) charged to Wellcome (inc VAT when charged)"].mode()

0    2040.0
dtype: float64

In [40]:
df["COST (£) charged to Wellcome (inc VAT when charged)"].median()

1884.01

In [41]:
df["COST (£) charged to Wellcome (inc VAT when charged)"].std()

146860.6655590476

In [44]:
df.groupby("Journal title").describe()

COST (£) charged to Wellcome (inc VAT when charged)  \
                                                                                                 count   
Journal title                                                                                            
academy of nutrition and dietetics                                                                1.0    
acs chemical biology                                                                              5.0    
acs chemical neuroscience                                                                         1.0    
acs nano                                                                                          2.0    
acta crystallographica section d,  biological c...                                                1.0    
acta crystallographica section d: biological cr...                                                1.0    
acta crystallographica section f: structural bi...                                                2.0    
acta crystallographica, section d                                                                 1.0    
acta crystallography d                                                                            1.0    
acta d                                                                                            1.0    
acta dermato venereologica                                                                        1.0    
acta diabetologica                                                                                1.0    
acta f                                                                                            1.0    
acta neuropathol                                                                                  1.0    
acta neuropathologica                                                                             3.0    
acta opthalmologica                                                                               1.0    
acta physiol                                                                                      1.0    
addiction                                                                                         2.0    
advances in experimental medicine and biology                                                     4.0    
age                                                                                               2.0    
age and ageing                                                                                    1.0    
ageing & society                                                                                  1.0    
aging cell                                                                                        1.0    
aids                                                                                              3.0    
aids behav                                                                                        1.0    
aids care                                                                                         4.0    
aids journal                                                                                      1.0    
aids research and therapy                                                                         1.0    
aids uk                                                                                           2.0    
alcohol and alcoholism                                                                            1.0    
...                                                                                               ...    
traffic                                                                                           6.0    
transactions of the royal society of tropical m...                                                1.0    
translational psychiatry                                                                          2.0    
trends in cognitive science                                                                       1.0    
trends in genetics     

The above chart shows (in the "mean" column) the price per article for a given journal, which is not exactly by subject.

In [67]:
df.groupby("Publisher").describe()

Article title         \
                                                           count unique   
Publisher                                                                 
ACS                                                            7      7   
ACS (Amercian Chemical Society) Publications                   5      5   
ACS Publications                                               2      2   
AGA Institute                                                  1      1   
AMBSB                                                          1      1   
AMERICAN CHEMICAL SOCIETY                                      1      1   
ASBMB                                                         46     46   
ASBMB Cadmus                                                   1      1   
ASBMB/Cadmus                                                   1      1   
ASBMB/Cenveo Publisher Services                                1      1   
ASBMC /CENVEO                                                  1      1   
ASM                                                            2      2   
ASM (American Society for Microbiology)                        1      1   
American Association of Immunologists                          2      2   
American Chemical Society                                     14     14   
American Chemical Society Publications                         1      1   
American College of Chest Physicians                           1      1   
American Physiological Society                                 7      7   
American Psychiatric Association                               1      1   
American Psychiatric Publishing                                1      1   
American Psychological Association                            13     13   
American Psychological Association                             1      1   
American Public Health Association                             1      1   
American Soc for Biochemistry and Molecular Bio...             1      1   
American Society for Biochemistry and Molecular...             1      1   
American Society for Biochemistry and Molecular...            18     18   
American Society for Biochemistry and Molecular...             1      1   
American Society for Investigative Pathology                   1      1   
American Society for Microbiology                             16     16   
American Society for Microbiology                              1      1   
...                                                          ...    ...   
The Royal College of Psychiatrists                             2      2   
The Royal Society                                              6      6   
The Sheridan Press                                             3      3   
The Sheridan Press                                             1      1   
The company of Biolgists                                       2      2   
The company of Biologists                                      1      1   
The royal Society                                              1      1   
Transcript Verlag                                              1      1   
University of the Basque Country Press                         1      1   
Wiley                                                        136    136   
Wiley                                                         12     12   
Wiley & Son                                                    1      1   
Wiley Blackwell                                                4      4   
Wiley Online Library                                           1      1   
Wiley Subscription Services                                    8      8   
Wiley Subscription Services  Inc.                             13     13   
Wiley Subscription Services Inc                                2      2   
Wiley Subscription Serviices Inc                               1      1   
Wiley VCH                                                      1      1   
Wiley-Blackwell                                               56     56   
W

The above chart groups the articles by publisher, which probably is closer to grouping by subject.

In [68]:
import re

journals = df["Journal title"]

subject = ["biology"]

for journal in journals:
    if re.search(journal, subject):
        print("found a match")
#df["Journal title"].apply(lambda x: re.findall("biology", x)
                          
#example = ["veterinary studies", "journal of biology", "biology now"]

#re.findall("biology", example)

TypeError: expected string or bytes-like object

In [69]:
import re

In [78]:
df["Journal title"].apply(lambda x: re.findall("biology", x))

TypeError: expected string or bytes-like object

In [83]:
df["Subject"] = ""

In [84]:
df.head()

PMID/PMCID Publisher           Journal title  \
0                    NaN       CUP  psychological medicine   
1             PMC3679557       ACS       biomacromolecules   
2  23043264  PMC3506128        ACS              j med chem   
3    23438330 PMC3646402       ACS              j med chem   
4   23438216 PMC3601604        ACS              j org chem   

                                       Article title  \
0  Reduced parahippocampal cortical thickness in ...   
1  Structural characterization of a Model Gram-ne...   
2  Fumaroylamino-4,5-epoxymorphinans and related ...   
3  Orvinols with mixed kappa/mu opioid receptor a...   
4  Regioselective opening of myo-inositol orthoes...   

  COST (£) charged to Wellcome (inc VAT when charged) Subject  
0                                              £0.00           
1                                           £2381.04           
2                                            £642.56           
3                                            £669.64           
4                                            £685.88

In [98]:
subj = "biology"
journal = "biology rocks"
new_column = []

if subj in df["Journal title"]:
    df["Subject"] = "biology"


In [107]:
print(df["Journal title"])

0                                  psychological medicine
1                                       biomacromolecules
2                                              j med chem
3                                              j med chem
4                                              j org chem
5                          journal of medicinal chemistry
6                            journal of proteome research
7                                               mol pharm
8                                    acs chemical biology
9                                    acs chemical biology
10           journal of chemical information and modeling
11                         journal of medicinal chemistry
12                         journal of medicinal chemistry
13                                           biochemistry
14                         journal of medicinal chemistry
15                                       gastroenterology
16                        journal of biological chemistry
17            

In [130]:
def subject(c):
    if "biology" in str(c["Journal title"]):
        return "biology"
    elif "chem" in str(c["Journal title"]):
        return "chemistry"
    elif "psych" in str(c["Journal title"]):
        return "psychology"
    elif "immunology" in str(c["Journal title"]):
        return "immunology"
    elif "physiology" in str(c["Journal title"]):
        return "physiology"
    elif "aids" in str(c["Journal title"]):
        return "AIDS"
    else:
        return "undefined"
    
df["Subject"] = df.apply(subject, axis=1)
    

In [131]:
if "chem" in "j med chem":
    print("yup")

yup


In [133]:
df.tail(n=10)

PMID/PMCID                      Publisher  \
2117                       PMC3773237                  Wolters Kluwer   
2118                       PMC3815011                  Wolters Kluwer   
2119                        PMC3805356                 Wolters Kluwer   
2120                             Epub                  Wolters Kluwer   
2121                        PMC3707567                 Wolters Kluwer   
2122                           2901593          Wolters Kluwer Health   
2123                           3748854         Wolters Kluwer Health    
2124                           3785148         Wolters Kluwer Health    
2125          PMCID:\n    PMC3647051\n  Wolters Kluwer N.V./Lippinott   
2126  PMID: 23846567  (Epub July 2013)                Wolters Kluwers   

                                          Journal title  \
2117                                            aids uk   
2118                                            aids uk   
2119  jaids journal of acquired immune deficiency sy...   
2120  journal of acquired immune deficiency syndroms...   
2121                                    journal of aids   
2122                               circulation research   
2123                                               aids   
2124                               pediatr infect dis j   
2125                                               aids   
2126    journal of acquired immune deficiency syndromes   

                                          Article title  \
2117   Sexual behaviour in a rural high HIV prevalence    
2118  HIV incidence and survival from age-specific s...   
2119  Accelerated biological ageing in HIV-infected ...   
2120  Reduction in early mortality on antiretroviral...   
2121  Risk factors for seropositivity to Kaposi sarc...   
2122  Mechanistic Links Between Na+ Channel (SCN5A) ...   
2123  Evaluation of an empiric risk screening score ...   
2124  Topical umbilical cord care for prevention of ...   
2125  Grassroots Community Organisations' Contributi...   
2126  A novel community health worker tool outperfor...   

     COST (£) charged to Wellcome (inc VAT when charged)    Subject  
2117                                           £1836.92        AIDS  
2118                                           £1836.92        AIDS  
2119                                           £1375.80        AIDS  
2120                                           £1836.92        AIDS  
2121                                           £2009.65        AIDS  
2122                                           £1334.15   undefined  
2123                                           £1834.77        AIDS  
2124                                           £1834.77   undefined  
2125                                           £2374.52        AIDS  
2126                                           £2034.75   undefined

In [134]:
df["Subject"].value_counts()

undefined     1663
chemistry      168
biology        139
psychology      82
immunology      33
physiology      23
AIDS            19
Name: Subject, dtype: int64

In [135]:
df.groupby("Subject").describe()

Article title         \
                   count unique   
Subject                           
AIDS                  19     19   
biology              139    139   
chemistry            168    168   
immunology            33     33   
physiology            23     23   
psychology            82     82   
undefined           1663   1662   

                                                                    \
                                                          top freq   
Subject                                                              
AIDS        HIV incidence and survival from age-specific s...    1   
biology     An experimental model for studying the biomech...    1   
chemistry   Regioselective opening of myo-inositol orthoes...    1   
immunology  Rapid early innate control of hepatitis C viru...    1   
physiology  Bimodal collagen fibril diameter distributions...    1   
psychology  COMT Val158Met Genotyp Determines the Directio...    1   
undefined   Exclusive breastfeeding, diarrhoel morbidity a...    2   

           COST (£) charged to Wellcome (inc VAT when charged)         \
                                                         count unique   
Subject                                                                 
AIDS                                                       19      15   
biology                                                   139     117   
chemistry                                                 168     144   
immunology                                                 33      31   
physiology                                                 23      23   
psychology                                                 82      58   
undefined                                                1663    1088   

                          Journal title         \
                 top freq         count unique   
Subject                                          
AIDS        £1836.92    3            19     10   
biology     £1939.95    6           139     61   
chemistry   £1800.00   13           168     60   
immunology  £3000.00    3            33     14   
physiology  £2110.15    1            23     11   
psychology  £3000.00    9            82     46   
undefined   £2040.00   84          1662    726   

                                                 PMID/PMCID         \
                                        top freq      count unique   
Subject                                                              
AIDS                              aids care    4         19     19   
biology                     current biology   11        123    123   
chemistry   journal of biological chemistry   53        164    163   
immunology   european journal of immunology    8         32     32   
physiology            journal of physiology    8         20     20   
psychology                 neuropsychologia    7         73     70   
undefined                          plos one  190       1497   1461   

                                          Publisher         \
                                 top freq     count unique   
Subject                                                      
AIDS        PMCID:\n    PMC3564972\n    1        19     10   
biology                   PMC3477309    1       139     41   
chemistry                 PMC3173209    2       168     45   
immunology         PMCID: PMC3298641    1        33     16   
physiology                   3764640    1        23     11   
psychology                PMC3529206    2        82     24   
undefined          Not yet available    7      1663    238   

                                                 
                                       top freq  
Subject                                          
AIDS                        Wolters Kluwer    7  
biology                           Elsevier   49  
chemistry                            ASBMB   44  
immunology                        Elsevier    9  
physiology  American Physiological Society    7  
psych

The above table is the closest I have come to finding the open access prices by subject area.